### PR TITLE
Upload Case Document

### PR DESCRIPTION
**Pull request by murilovmachado**
Jira: https://bugs.benetech.org/browse/CHI-744
Primary reviwer: @GPaoloni

### CODE REVIEW

File: plugin-hrm-form/src/services/ServerlessService.ts [hunk 1]
```diff
@@ -137,3 +137,38 @@ export const getWorkerAttributes = async (workerSid: string) => {
   const response = await fetchProtectedApi('/getWorkerAttributes', body);
   return response;
 };
+
+const toBase64 = file =>
```
Comment (nickhurlburt) [inline plugin-hrm-form/src/services/ServerlessService.ts:141]: Is this still in use?

**Review by nickhurlburt — APPROVED** [review]
I gave this a very quick review.  Looks good to me.  I don't see any security issues since all of that is handled in serverless functions, which is good.

There are a couple things that I've realized are O(N) in the number of helplines we support, whenever we add a new feature like this:
1. Adding a new JSON file to each configuration (like `DocumentForm.json`)
2. Adding translations for new strings (not an issue now because it's still all English)

Nothing to do about that right now, but when these start to become heavier tasks, we will want to invest in making these easier to support as N gets bigger.

File: plugin-hrm-form/src/services/ServerlessService.ts (hunk 1 above)
Comment (murilovmachado) [reply plugin-hrm-form/src/services/ServerlessService.ts]: Good catch, removed.

File: plugin-hrm-form/src/components/case/AddDocument.tsx [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    const onDeleteFile = async (fileName: string) => {
+      await deleteFile(fileName);
+    };
+
+    const generatedForm = createFormFromDefinition(DocumentForm)([])(initialForm)(
+      updateCallBack,
+      onFileChange,
+      onDeleteFile,
+    );
```
Comment (murilovmachado) [inline plugin-hrm-form/src/components/case/AddDocument.tsx:124]: Here's probably the more important tech debt of this PR.

Notice that the way it's currently implemented, if we have multiple fields of type `file-upload`, (e.g.: `document1` and `document2`), the **upload** and **delete** behavior of both documents will be the same.

Example: We can't have `document1` to be uploaded to AWS S3, but `document2` to a different place. Currently they share the upload and delete functions.

**Comment by nickhurlburt** [comment]
One thing I forgot to ask: accessibility.  Does this decrease or increase our accessibility debt, or keep it the same?  Are actionable elements focusable, and non-actionable elements not focusable?  Do the focus for new views follow a sensible sequence?

The task to show focus is not until later, but this should be built so that focus works correctly when we start to show focus.

File: plugin-hrm-form/src/components/case/AddDocument.tsx (hunk 2 above)
Comment (nickhurlburt) [reply plugin-hrm-form/src/components/case/AddDocument.tsx:124]: This seems reasonable to me.

**Comment by murilovmachado** [comment]
> One thing I forgot to ask: accessibility. Does this decrease or increase our accessibility debt, or keep it the same? Are actionable elements focusable, and non-actionable elements not focusable? Do the focus for new views follow a sensible sequence?
> 
> The task to show focus is not until later, but this should be built so that focus works correctly when we start to show focus.

The thing that can impact accessibility the most on this PR is the `UploadFileInput` component. The native `<input type="file">` is very poorly customizable, and could not meet our designs at all. So, instead, we're using a `<input type="button">` which the user interacts with, that basically forwards its click to a **hidden** `<input type="file">`. I think this is a good topic to discuss with James/Sina, and see what they have to say about this component.

**Review by GPaoloni — APPROVED** [review]
Code looks great, behavior is the expected. Really great job!! :rocket: 

Just one minor UI bug. When you try to save with no document, the error message pops up indicating the issue.Then I added a document but the UI didn't notice this change and the form state remains in error, not being able to save the document in HRM (even if it's already saved in AWS).
Screenshot from 2021-09-14 16-20-40

File: plugin-hrm-form/src/formDefinitions/et-v1/caseForms/DocumentForm.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+[
```
[the remark this answers is not in the document]
Comment (murilovmachado) [reply plugin-hrm-form/src/formDefinitions/et-v1/caseForms/DocumentForm.json:1]: Yes, similar to other forms, we have `const document = methods.getValues();` to get all values defined in the form. So, if the form definitions have more fields, all of them will be saved in the `document` object.

File: plugin-hrm-form/src/components/common/forms/formGenerators.tsx [hunk 4]
```diff
@@ -91,8 +95,12 @@ const bindCreateSelectOptions = (path: string) => (o: SelectOption) => (
  * @param {() => void} updateCallback Callback called to update form state. When is the callback called is specified in the input type.
  * @param {FormItemDefinition} def Definition for a single input.
  */
-// eslint-disable-next-line sonarjs/cognitive-complexity
-const getInputType = (parents: string[], updateCallback: () => void) => (def: FormItemDefinition) => (
+const getInputType = (
+  parents: string[],
+  updateCallback: () => void,
+  onFileChange?: (event: any) => Promise<string>,
+  onDeleteFile?: (fileName: string) => Promise<void>,
```
[the remark this answers is not in the document]
Comment (murilovmachado) [reply plugin-hrm-form/src/components/common/forms/formGenerators.tsx]: Great idea, thanks!

**Pull request merged**